src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -364,34 +372,406 @@ public void WriteValue(sbyte value)
 
         public void WriteValue(string value)
         {
-            this.WriteSeparatorIfNecessary();
             if (value == null)
             {
+                this.WriteSeparatorIfNecessary();
                 this.utf8JsonWriter.WriteNullValue();
             }
+            else if (value.Length > bufferWriter.FreeCapacity)

```

Comment:
can we do 3/4?

Reply:
Maybe we should place an upper bound, like:

```c#
if (value.Length > bufferWriter.FreeCapacity || value.Length > someThreshold
```
Where that `someThreshold` could be something like 12k if we are okay with some buffer resize but want to avoid LOH. Or it could be something small like 2048 if we want to avoid buffer resize more aggressively.

Reply:
I set a threshold of 2048.